src/rust/ide/json-rpc/src/macros.rs

```diff
@@ -0,0 +1,129 @@
+//! Helper macros to generate RemoteClient and MockClient.
+
+// FIXME[dg]: https://github.com/luna/ide/issues/401 We want to make the generated methods to
+// take references instead of ownership.
+/// This macro reads a `trait API` item and generates asynchronous methods for RPCs. Each method
+/// should be signed with `MethodInput`, `rpc_name`, `result` and `set_result` attributes. e.g.:
+/// ```rust,compile_fail
+/// make_rpc_method!{
+///     trait API {
+///         #[MethodInput=CallMePleaseInput,camelCase=callMePlease,result=call_me_please_result,
+///         set_result=set_call_me_please_result]
+///         fn call_me_please(&self, my_number_is:String) -> ();
+///     }
+/// }
+/// ```
+///
+/// This macro generates an `API` trait and creates two structs implementing `API`
+/// called `Client`, with the actual RPC methods, and `MockClient`, with mocked methods with
+/// return types setup by:
+/// ```rust,compile_fail
+///     fn set_call_me_please_result
+///     (&mut self, my_number_is:String,result:json_rpc::api::Result<()>) { /* impl */ }
+/// ```
+#[macro_export]
+macro_rules! make_rpc_methods {
+    (
+    $(#[doc = $impl_doc:expr])+
+    trait API {
+        $($(#[doc = $doc:expr])+
+        #[MethodInput=$method_input:ident,rpc_name=$rpc_name:expr,result=$method_result:ident,
+        set_result=$set_result:ident]
+        fn $method:ident(&self $(,$param_name:ident:$param_ty:ty)+) -> $result:ty;
+        )*
+    }) => {
+        // ===========
+        // === API ===
+        // ===========
+
+        $(#[doc = $impl_doc])+
+        pub trait API {
+            $($(#[doc = $doc])+
```

Comment:
make newline and indentation after first `$(` because you are not closing it in the same line